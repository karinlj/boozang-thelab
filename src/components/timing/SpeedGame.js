import { useState, useEffect } from "react";
import "./timing.scss";
import { SpeedGameVideos } from "../text/videos/VideoSections";
import { SpeedGameIntro } from "../text/Intros";
import { SpeedGameTestInfo } from "../text/TestInfos";
import ResultMessages from "../compMessages/ResultMessages";

const SpeedGame = () => {
  const [isRunning, setIsRunning] = useState(false);
  const [count, setCount] = useState(0);
  const [successMessage, setSuccessMessage] = useState("");
  const [timeMessage, setTimeMessage] = useState("");

  const handleStart = () => {
    if (!isRunning) {
      setIsRunning(true);
      setSuccessMessage("");
      setTimeMessage("");
      let startCount = Math.floor(Math.random() * 10000) + 1000; //random nr mellan 1000-10000ms
      //start counting down at randomnr
      setCount(startCount);
    }
  };
  const handleStop = () => {
    if (isRunning) {
      setIsRunning(false);
      setSuccessMessage("You are fast!!");
      setTimeMessage(`Your reaction time is ${-count} ms.`);
    }
  };
  useEffect(() => {
    let myInterval = null;
    if (isRunning) {
      myInterval = setInterval(() => {
        setCount((count) => count - 100);
      }, 100);
    } else if (!isRunning) {
      clearInterval(myInterval);
    }
    //cleanup
    return () => clearInterval(myInterval);
  }, [isRunning, count]);

  return (
    <div className="row justify-content-between" data-testid="SpeedGame">
      <div className="col-12 col-md-6 col-xl-5">
        <section className="timing_section">
          <SpeedGameIntro />
          <div className="game_space">
            <input
              type="button"
              value="Start Game"
              className="formBtn add"
              onClick={handleStart}
              data-testid="startBtn"
            />
            {count < 0 && (
              <input
                type="button"
                value="End Game"
                className="formBtn delete"
                onClick={handleStop}
                data-testid="stopBtn"
              />
            )}
            <ResultMessages
              openWrapper={!isRunning}
              successMessage={successMessage}
              subMessage={timeMessage}
              messageString={"You are fast!!"}
            />
          </div>
        </section>
      </div>
      <div className="col-12 col-md-5">
        <SpeedGameTestInfo />
        <SpeedGameVideos />
      </div>
    </div>
  );
};
export default SpeedGame;
