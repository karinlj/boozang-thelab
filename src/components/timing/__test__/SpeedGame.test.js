import SpeedGame from "../SpeedGame";
import { render, fireEvent, screen, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

afterEach(() => {
  cleanup();
});
// test("should render SpeedGame component", () => {
//   render(<SpeedGame />);
//   const SpeedGameElement = screen.getByTestId("SpeedGame");
//   expect(SpeedGameElement).toBeInTheDocument();
// });

// test("should render SpeedGame component", () => {
//   const tree = renderer.create(<SpeedGame />).toJSON();
//   expect(tree).toMatchSnapshot();
// });

test("resultMessage should have correct classname when click on Start", () => {
  //   const { debug } = render(<SpeedGame />);
  //   debug();
  const { getByTestId } = render(<SpeedGame test="0" />);
  const startBtnElement = getByTestId("startBtn");

  const resultMessage = getByTestId("result");

  fireEvent.click(startBtnElement);

  expect(resultMessage).toHaveClass("result_wrapper");
});

// test("function handleStart should be called when click on startBtn", () => {
//mock a function
//   const handleStart = jest.fn();
//   const { getByTestId } = render(<SpeedGame />);
// });

// test("test", () => {
//   expect(true).toBe(true);
// });
