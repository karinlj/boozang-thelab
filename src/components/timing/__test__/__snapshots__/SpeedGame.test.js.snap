// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SpeedGame component 1`] = `
<div
  className="row justify-content-between"
  data-testid="SpeedGame"
>
  <div
    className="col-12 col-md-6 col-xl-5"
  >
    <section
      className="timing_section"
    >
      <article
        className="comp_intro"
      >
        <h1>
          Speed Game
        </h1>
        <p>
          This game tests the user reaction time. This is also useful to show-case test automation render waits. where x is a random time between 1 and 10 seconds. The hit the button as fast as you can. The smaller the number of milliseconds above, the better.
        </p>
      </article>
      <div
        className="game_space"
      >
        <input
          className="formBtn add"
          onClick={[Function]}
          type="button"
          value="Start Game"
        />
        <div
          className="result_wrapper show"
        >
          <h4
            className="success_message fail"
          >
            
          </h4>
          
        </div>
      </div>
    </section>
  </div>
  <div
    className="col-12 col-md-5"
  >
    <article
      className="comp_test_info"
    >
      <h4>
        Why learn: Speed Game
      </h4>
      <p>
        A classic problem in test automation is handling render waits. If certain elements are taking longer to render, the test might fail. On the other hand, if we create long delays the overall test takes too long to run. Best if is we can wait until an element appears, but then click it as fast as we can.
      </p>
    </article>
    <section
      className="video_section"
    >
      <div
        className=""
      >
        <div>
          <h5>
            Related videos on YouTube
          </h5>
          <ul>
            <li>
              <a
                href="https://www.youtube.com/watch?v=cex5hf5ui6M&list=PLhdB7gylbUzpRBVLjpi4KI-g2j7U0y9Qr&index=5"
                rel="noopener noreferrer"
                target="_blank"
              >
                <img
                  alt="YouTube logo"
                  className="youtube_logo"
                  src="youtube_logo.png"
                />
                <span>
                   
                  Speed
                </span>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </section>
  </div>
</div>
`;
